gpflow/models/util.py

```diff
@@ -52,3 +54,18 @@ def maximum_log_likelihood_objective(model: BayesianModel, data: Data) -> tf.Ten
     else:
         _assert_equal_data(model.data, data)
         return model.maximum_log_likelihood_objective()
+
+
+def data_input_to_tensor(structure):
+    """
+    Converts non-tensor elements of a structure to TensorFlow tensors retaining the structure itself.
+    The function doesn't keep original element's dtype and forcefully converts
+    them to GPflow's default float type.
+    """
+
+    def convert_to_tensor(elem):
+        if tf.is_tensor(elem) or isinstance(elem, Iterator):
+            return elem
+        return tf.convert_to_tensor(elem, dtype=default_float())
```

Comment:
Why not keep the original dtype?

Reply:
Because it is not possible to deduce the original type in python for sequence structures.